{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import seaborn as sns\n",
    "sns.set(font_scale=1)\n",
    "from plotnine import *\n",
    "from tqdm.notebook import tqdm\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from pandarallel import pandarallel\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "pd.set_option(\"display.max_columns\", None)\n",
    "pd.set_option('display.width', 1000)\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "PAIRED_DATASETS = pd.read_csv(\"/home/pbarbosa/git_repos/mutsplice/data/2_paired_datasets/ALL_data.tsv.gz\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-03-26 15:40:59.019448: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-03-26 15:41:00.224933: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory\n",
      "2024-03-26 15:41:00.225014: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory\n",
      "2024-03-26 15:41:00.225021: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    }
   ],
   "source": [
    "from gtfhandle.utils import file_to_bed_df\n",
    "from mutsplice.datasets.mutsplice_pipeline import MutSplicePipeline\n",
    "import os\n",
    "\n",
    "def runMutSplice(df, rbp_name: str, exon_group: str, motif_source: str, motif_search: str):\n",
    "    GTF_CACHE = \"/home/pbarbosa/data/genomes/hg38/gtf_cache_gencode/\"\n",
    "    OUT_DIR = f\"/big_data/pbarbosa/mutsplice/{motif_source}_{motif_search}/\"\n",
    "    FASTA = \"/home/pbarbosa/data/genomes/hg38/GRCh38.primary_assembly.genome.fa\"\n",
    "\n",
    "    kwargs = {\n",
    "        \"gtf_cache\": GTF_CACHE,\n",
    "        \"fasta\": FASTA,\n",
    "        \"out_dir\": f\"{OUT_DIR}{rbp_name}_{exon_group}\",\n",
    "        \"outbasename\": f\"{rbp_name}\",\n",
    "        \"subset_rbps\": [rbp_name],\n",
    "        \"motif_source\": motif_source,\n",
    "        \"motif_search\": motif_search,\n",
    "        \"pvalue_threshold\": 0.0005,\n",
    "        \"spliceai_save_raw\": True,\n",
    "        \"spliceai_raw_results\": None,\n",
    "        \"spliceai_final_results\": None,\n",
    "        \"batch_size\": 64,\n",
    "        \"use_full_sequence\": False,\n",
    "    }\n",
    "    bed_df = file_to_bed_df(df, is_0_based=False, header=0, col_index=0)\n",
    "\n",
    "    MutSplicePipeline(\n",
    "        bed_df,\n",
    "        do_gtf_queries=True,\n",
    "        do_motif_scanning=True,\n",
    "        do_mutations=True,\n",
    "        run_spliceai=True,\n",
    "        **kwargs,\n",
    "    )\n",
    "\n",
    "\n",
    "def process_single_RBP(group, exon_group: str, motif_source: str, motif_search: str):\n",
    "    try:\n",
    "        rbp_name = group.iloc[0].rbp_name\n",
    "        runMutSplice(\n",
    "            group,\n",
    "            rbp_name=rbp_name,\n",
    "            exon_group=exon_group,\n",
    "            motif_source=motif_source,\n",
    "            motif_search=motif_search,\n",
    "        )\n",
    "\n",
    "    except ValueError as ve:\n",
    "        print(f\"Caught ValueError for {rbp_name}: {ve}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Knockdown exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_coordinates</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>rbp_name</th>\n",
       "      <th>exon_group</th>\n",
       "      <th>dPSI</th>\n",
       "      <th>paired_with</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2:130161186-130161272</td>\n",
       "      <td>-</td>\n",
       "      <td>SMPD4</td>\n",
       "      <td>ENST00000680298</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>KD</td>\n",
       "      <td>0.229</td>\n",
       "      <td>chr22:35338713-35338788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:47259180-47259259</td>\n",
       "      <td>+</td>\n",
       "      <td>NR1H3</td>\n",
       "      <td>ENST00000441012</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>KD</td>\n",
       "      <td>0.442</td>\n",
       "      <td>chr20:37401602-37401678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr20:49850810-49850844</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC9A8</td>\n",
       "      <td>ENST00000361573</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>KD</td>\n",
       "      <td>0.612</td>\n",
       "      <td>chr15:84679932-84679969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9:96852853-96853069</td>\n",
       "      <td>-</td>\n",
       "      <td>ZNF782</td>\n",
       "      <td>ENST00000481138</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>KD</td>\n",
       "      <td>0.576</td>\n",
       "      <td>chr15:72349079-72349259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr9:133465322-133465447</td>\n",
       "      <td>+</td>\n",
       "      <td>CACFD1</td>\n",
       "      <td>ENST00000316948</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>KD</td>\n",
       "      <td>-0.273</td>\n",
       "      <td>chr7:857828-857957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30460</th>\n",
       "      <td>chr14:91145221-91145313</td>\n",
       "      <td>+</td>\n",
       "      <td>DGLUCY</td>\n",
       "      <td>ENST00000518868</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>KD</td>\n",
       "      <td>-0.598</td>\n",
       "      <td>chr16:28581055-28581144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30462</th>\n",
       "      <td>chr12:114679913-114679972</td>\n",
       "      <td>-</td>\n",
       "      <td>TBX3</td>\n",
       "      <td>ENST00000257566</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>KD</td>\n",
       "      <td>-0.174</td>\n",
       "      <td>chr11:62571838-62571901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30464</th>\n",
       "      <td>chr10:96954953-96955834</td>\n",
       "      <td>+</td>\n",
       "      <td>LCOR</td>\n",
       "      <td>ENST00000540664</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>KD</td>\n",
       "      <td>0.419</td>\n",
       "      <td>chr2:61214060-61214694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30466</th>\n",
       "      <td>chr10:96954953-96955671</td>\n",
       "      <td>+</td>\n",
       "      <td>LCOR</td>\n",
       "      <td>ENST00000676381</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>KD</td>\n",
       "      <td>0.411</td>\n",
       "      <td>chr2:23757418-23758101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30468</th>\n",
       "      <td>chr10:96954953-96955454</td>\n",
       "      <td>+</td>\n",
       "      <td>LCOR</td>\n",
       "      <td>ENST00000676414</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>KD</td>\n",
       "      <td>0.419</td>\n",
       "      <td>chr2:256207-256690</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15235 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              target_coordinates Strand gene_name    transcript_id rbp_name exon_group   dPSI              paired_with\n",
       "0       chr2:130161186-130161272      -     SMPD4  ENST00000680298     ADAR         KD  0.229  chr22:35338713-35338788\n",
       "2        chr11:47259180-47259259      +     NR1H3  ENST00000441012     ADAR         KD  0.442  chr20:37401602-37401678\n",
       "4        chr20:49850810-49850844      +    SLC9A8  ENST00000361573     ADAR         KD  0.612  chr15:84679932-84679969\n",
       "6         chr9:96852853-96853069      -    ZNF782  ENST00000481138     ADAR         KD  0.576  chr15:72349079-72349259\n",
       "8       chr9:133465322-133465447      +    CACFD1  ENST00000316948     ADAR         KD -0.273       chr7:857828-857957\n",
       "...                          ...    ...       ...              ...      ...        ...    ...                      ...\n",
       "30460    chr14:91145221-91145313      +    DGLUCY  ENST00000518868   ZRANB2         KD -0.598  chr16:28581055-28581144\n",
       "30462  chr12:114679913-114679972      -      TBX3  ENST00000257566   ZRANB2         KD -0.174  chr11:62571838-62571901\n",
       "30464    chr10:96954953-96955834      +      LCOR  ENST00000540664   ZRANB2         KD  0.419   chr2:61214060-61214694\n",
       "30466    chr10:96954953-96955671      +      LCOR  ENST00000676381   ZRANB2         KD  0.411   chr2:23757418-23758101\n",
       "30468    chr10:96954953-96955454      +      LCOR  ENST00000676414   ZRANB2         KD  0.419       chr2:256207-256690\n",
       "\n",
       "[15235 rows x 8 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KD_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 1 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "pandarallel.initialize(progress_bar=True, nb_workers=1)\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"0\"\n",
    "\n",
    "KD_data = PAIRED_DATASETS[PAIRED_DATASETS.exon_group == \"KD\"]\n",
    "# KD_data.groupby(\"rbp_name\").parallel_apply(\n",
    "#     process_single_RBP, exon_group=\"KD\", motif_source=\"ATtRACT\", motif_search=\"fimo\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Control exons "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pandarallel.initialize(progress_bar=True, nb_workers=1)\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"0\"\n",
    "\n",
    "Ctrl_data = PAIRED_DATASETS[PAIRED_DATASETS.exon_group == \"CTRL\"]\n",
    "Ctrl_data.groupby(\"rbp_name\").parallel_apply(\n",
    "    process_single_RBP,\n",
    "    exon_group=\"CTRL\",\n",
    "    motif_source=\"ATtRACT\",\n",
    "    motif_search=\"fimo\",\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mutsplice",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
